File: nim/binary-search/binary_search.nim
proc binarySearch*(a: openArray[int], val: int): int =
  var left = 0
  var right = len a
  while left < right:
    var mid = left + (right - left) div 2
    if a[mid] == val: return mid
    elif a[mid] < val: left = mid + 1
    else: right = mid
  return -1

File: c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
   size_t left = 0;
   size_t right = length;

   while (left < right) {
      size_t mid = (left + right) / 2;
      int mid_value = arr[mid];

      if (mid_value < value) {
         left = mid + 1;
      } else if (mid_value > value) {
         right = mid;
      } else {
         return &arr[mid];
      }
   }
   return NULL;
}
